### PR TITLE
CDI: Add `mount-nvidia-binaries` and `mount-nvidia-docker-1-directories` options

### PR DESCRIPTION
**Pull request by ereslibre**
Add three options to `hardware.nvidia-container-toolkit`:

- `mounts`: list of  mounts that allow to mount arbitrary paths on the CDI enabled containers.

- `mount-nvidia-binaries`: this option allows users to avoid mounting nvidia binaries on the container.

- `mount-nvidia-docker-1-directories`: this option allows users to avoid mounting `/usr/local/nvidia/lib{,64}` on containers.

Related: https://github.com/NixOS/nixpkgs/issues/290609

## Things done

I have tested that `podman run --device=nvidia.com/gpu=all -d -p 11434:11434 --name ollama ollama/ollama` works as expected with this change, and is able to use all the GPU's in my system without any change.

My NixOS configuration sets `virtualisation.containers.cdi.dynamic.nvidia.enable = true;`.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix [hunk 1]
```diff
@@ -11,6 +11,12 @@
       containerPath = "/usr/bin/nvidia-powerd"; }
     { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-smi";
       containerPath = "/usr/bin/nvidia-smi"; }
+    # nvidia-docker 1.0 uses /usr/local/nvidia/lib{,64} on their LD_LIBRARY_PATH
+    #   e.g. https://gitlab.com/nvidia/container-images/cuda/-/blob/e3ff10eab3a1424fe394899df0e0f8ca5a410f0f/dist/12.3.1/ubi9/base/Dockerfile#L44
+    { hostPath = "${config.hardware.nvidia.package}/lib";
+      containerPath = "/usr/local/nvidia/lib"; }
+    { hostPath = "${config.hardware.nvidia.package}/lib";
+      containerPath = "/usr/local/nvidia/lib64"; }
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:19]: OK, so we're looking to institutionalize one more upstream hack that is broken by design and was never meant to work. Perhaps. Can we find a reference for the runtime wrapper actually populating these paths? Are we sure these paths aren't populated by the `yum install` from their `Dockerfile`?

Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:19]: It's not about institutionalizing upstream hacks, in my opinion it's a matter of having a reasonable expectation of running container images that are already published in multiple places without having to worry about them as a user.

Why, as a NixOS user, do I have to worry about `podman run ollama/ollama` giving me an obscure error to learn that I need to tweak the `LD_LIBRARY_PATH` pointing to some `/nix/store` that changes over time? In my opinion that sounds miserable as a user experience.

Comment (SomeoneSerge) [reply nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:19]: If upstream produces a broken image, the preference is to contact upstream and unbreak the image. This diff makes me suspect they might be building broken images (unless they do populate the path themselves in that `yum install`). Also it's extremely fishy that _you_ would need that `LD_LIBRARY_PATH` to work because it does not expose absolutely anything that you haven't exposed via `ld.so.cache` already.

TLDR: I'd like to understand what's going on a bit better

Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:19]: > Can we find a reference for the runtime wrapper actually populating these paths? Are we sure these paths aren't populated by the yum install from their Dockerfile?

https://github.com/dalg24/nvidia-docker/blob/c68a91358e737e33aa949b45c8ad069ae2cac9b6/src/nvidia/volumes.go#L107

**Comment by ereslibre** [comment]
> If upstream produces a broken image

They are not producing a broken image. When you use CUDA in container images it's common practice to do this. With the nvidia-docker, you would:

1. Set LD_LIBRARY_PATH to /usr/local/nvidia/lib{,64}
2. Something --a runtime wrapper,likely-- mounts the required drivers and libraries from the host at /usr/local/nvidia/lib{,64}
3. Your container runtime hopefully runs

Later on, I don't know exactly the timeline or when that happened, they changed the story a little bit:

1. You generate a CDI spec, that provides you with the ability to extend OCI hooks and mount other directories from the host.
2. With CDI, you mount the relevant drivers and binaries from the host on the guest.
3. An OCI hook, injected by CDI, takes the /etc/ld.so.cache on the container, and updates it, by adding the discovered libcuda.so.* and friends.
4. Your container starts, and hopefully runs

It's not ideal, it's probably not the most stable thing, but it's that we have. Many, many, many container images use this as of today. Our interest as NixOS developers is to find the balance between a perfect implementation, and letting people run their stuff without getting in their way as much as possible.

**Comment by SomeoneSerge** [comment]
If they actually enforce that these paths exist, they are broken:) If they were not broken, their `LD_LIBRARY_PATH` would be a no-op and you'd not get an error because the `ld.so` would still look up the `ld.so.cache`

**Comment by SomeoneSerge** [comment]
> They are not broken, again. They just took the decision that the container images would place the LD_LIBRARY_PATH at the place where the runtime wrapper will mount the host libraries (or ld.so.conf.d being augmented with it) --see official example of creating a cuda image from nvidia:--

The `LD_LIBRARY_PATH` set up by the container shouldn't matter at all, we have already informed the dynamic loader about the drivers' location through other means

**Comment by ereslibre** [comment]
> The LD_LIBRARY_PATH set up by the container shouldn't matter at all, we have already informed the dynamic loader about the drivers' location through other means

For the context of this PR, it doesn't matter whether it's LD_LIBRARY_PATH or ld.so.conf.d. Side note: I think the applications depending on CUDA tend to do a dlopen, that is affected by `LD_LIBRARY_PATH`. I am not so sure that it should not matter at all, I tend to think otherwise.

In any case, I did show both:

1. How a CUDA container image should be built according to nvidia:
    1. Setting LD_LIBRARY_PATH
    2. Modifying /etc/ld.so.conf.d
2. nvidia-docker doing this very same thing

**Comment by SomeoneSerge** [comment]
> For the context of this PR, it doesn't matter whether it's LD_LIBRARY_PATH or ld.so.conf.d. Side note: I think the applications depending on CUDA tend to do a dlopen, that is affected by LD_LIBRARY_PATH. I am not so sure that it should not matter at all, I tend to think otherwise.

The `dlopen` should work out of the box regardless of `LD_LIBRARY_PATH`, because our CDI hook already takes care of this.

Does your example (ollama) actually fail without these extra mounts?

**Comment by ereslibre** [comment]
> The dlopen should work out of the box regardless of LD_LIBRARY_PATH, because our CDI hook already takes care of this.

No, if the code in the container does a `dlopen` it's already on its own. Only LD_LIBRARY_PATH will matter.

**Comment by SomeoneSerge** [comment]
>      No, if the code in the container does a dlopen it's already on its own. Only LD_LIBRARY_PATH will matter.

No, `dlopen` is handled by the dynamic loader, which follows exactly the same logic as `DT_NEEDED`: it'll see if the library is already loaded (e.g. by `LD_PRELOAD`), it'll consult `LD_LIBRARY_PATH`, it'll consult `DT_RUNPATH` of the calling process, it'll consult the `ld.so.cache`s

**Comment by ereslibre** [comment]
> No, dlopen is handled by the dynamic loader, which follows exactly the same logic as DT_NEEDED: it'll see if the library is already loaded (e.g. by LD_PRELOAD), it'll consult LD_LIBRARY_PATH, it'll consult DT_RUNPATH of the calling process, it'll consult the ld.so.caches

Yes, you are right, my bad. In any case, we still need to mount this directory.

**Comment by SomeoneSerge** [comment]
> Yes, you are right, my bad. In any case, we still need to mount this directory.

Why, have we seen any failures?

**Comment by ereslibre** [comment]
As per PR description: `podman run --device=nvidia.com/gpu=all -d -p 11434:11434 --name ollama ollama/ollama` does not work with GPU acceleration without  this PR.

**Comment by SomeoneSerge** [comment]
Very good. Could you publish the logs so we can see what the error was?

**Comment by ereslibre** [comment]
> Very good, and sorry I missed this. Could you publish the logs so we can see what the error was?

```
❯ podman run --rm --device=nvidia.com/gpu=all -e OLLAMA_DEBUG="1" -p 11434:11434 --name ollama ollama/ollama serve
Couldn't find '/root/.ollama/id_ed25519'. Generating new private key.
Your new public key is:

ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAaCxFGYXlp+0XG39EwoWKRXrCqlhbn4rqYwO7d9OdY8

time=2024-02-24T10:50:18.777Z level=INFO source=images.go:710 msg="total blobs: 0"
time=2024-02-24T10:50:18.777Z level=INFO source=images.go:717 msg="total unused blobs removed: 0"
time=2024-02-24T10:50:18.777Z level=INFO source=routes.go:1019 msg="Listening on [::]:11434 (version 0.1.27)"
time=2024-02-24T10:50:18.777Z level=INFO source=payload_common.go:107 msg="Extracting dynamic libraries..."
time=2024-02-24T10:50:20.600Z level=INFO source=payload_common.go:146 msg="Dynamic LLM libraries [cuda_v11 cpu cpu_avx cpu_avx2 rocm_v6 rocm_v5]"
time=2024-02-24T10:50:20.601Z level=DEBUG source=payload_common.go:147 msg="Override detection logic by setting OLLAMA_LLM_LIBRARY"
time=2024-02-24T10:50:20.601Z level=INFO source=gpu.go:94 msg="Detecting GPU type"
time=2024-02-24T10:50:20.601Z level=INFO source=gpu.go:265 msg="Searching for GPU management library libnvidia-ml.so"
time=2024-02-24T10:50:20.601Z level=DEBUG source=gpu.go:283 msg="gpu management search paths: [/usr/local/cuda/lib64/libnvidia-ml.so* /usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-ml.so* /usr/lib/x86_64-linux-gnu/libnvidia-ml.so* /usr/lib/wsl/lib/libnvidia-ml.so* /usr/lib/wsl/drivers/*/libnvidia-ml.so* /opt/cuda/lib64/libnvidia-ml.so* /usr/lib*/libnvidia-ml.so* /usr/local/lib*/libnvidia-ml.so* /usr/lib/aarch64-linux-gnu/nvidia/current/libnvidia-ml.so* /usr/lib/aarch64-linux-gnu/libnvidia-ml.so* /opt/cuda/targets/x86_64-linux/lib/stubs/libnvidia-ml.so* /usr/local/nvidia/lib/libnvidia-ml.so* /usr/local/nvidia/lib64/libnvidia-ml.so*]"
time=2024-02-24T10:50:20.601Z level=INFO source=gpu.go:311 msg="Discovered GPU libraries: []"
time=2024-02-24T10:50:20.601Z level=INFO source=gpu.go:265 msg="Searching for GPU management library librocm_smi64.so"
time=2024-02-24T10:50:20.601Z level=DEBUG source=gpu.go:283 msg="gpu management search paths: [/opt/rocm*/lib*/librocm_smi64.so* /usr/local/nvidia/lib/librocm_smi64.so* /usr/local/nvidia/lib64/librocm_smi64.so*]"
time=2024-02-24T10:50:20.601Z level=INFO source=gpu.go:311 msg="Discovered GPU libraries: []"
time=2024-02-24T10:50:20.601Z level=INFO source=cpu_common.go:11 msg="CPU has AVX2"
time=2024-02-24T10:50:20.601Z level=DEBUG source=amd.go:32 msg="amd driver not detected /sys/module/amdgpu"
time=2024-02-24T10:50:20.601Z level=INFO source=routes.go:1042 msg="no GPU detected"
```

**Comment by ereslibre** [comment]
So yeah this instance is pretty hardcoded in ollama, what we could fix to make it a better citizen to be honest: https://github.com/ollama/ollama/blob/2a4b128ae3e3a18b10e8701aca2434d401eaa7ba/gpu/gpu.go#L37-L51

But honestly, I think this is a very common practice I fear.

**Comment by SomeoneSerge** [comment]
Aha. I'd need to dig deeper, but looking at https://github.com/ollama/ollama/blob/2a4b128ae3e3a18b10e8701aca2434d401eaa7ba/gpu/gpu.go#L38-L50, it's more than likely they do not use `dlopen` with library names (as they would if they actually targeted nvidia-docker), but hard-code the absolute paths. So there's a possibility we're not even talking about institutionalizing nvidia's broken designs, but about accommodating a single project that fails to integrate with these nvidia's designs

**Comment by ereslibre** [comment]
@SomeoneSerge we basically saw the same thing at the same time and wrote the same thing hehe.

I would settle it like this: it makes some broken published software work, and the old wrapper was doing it as well; also, we don't lose much in the process.

**Comment by ereslibre** [comment]
I have to say: and thanks for pushing to find the real root cause :)

**Comment by SomeoneSerge** [comment]
Ah, you were faster:)

> But honestly, I think this is a very common practice I fear.

What I'm thinking is, and this concerns the `/usr/bin` and `/usr/lib` paths from the original PR, we could add options like `cdi.dynamic.nvidia.mount-usr-local`, `mount-usr-bin`, etc.

**Comment by ereslibre** [comment]
> What I'm thinking is, and this concerns the /usr/bin and /usr/lib paths from the original PR, we could add options like cdi.dynamic.nvidia.mount-usr-local, mount-usr-bin, etc. I'd set them to false by default

Would you mind if we discuss that at https://github.com/NixOS/nixpkgs/issues/290609? I would like to discuss that a bit more in detail and understand better what you mean by that.

File: nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:1]: > > What I'm thinking is, and this concerns the /usr/bin and /usr/lib paths from the original PR, we could add options like cdi.dynamic.nvidia.mount-usr-local, mount-usr-bin, etc. I'd set them to false by default
> 
> Would you mind if we discuss that at #290609? I would like to discuss that a bit more in detail and understand better what you mean by that.

Actually, I'd prefer to contain this discussion to a thread, so we can attach it to this PR and link it elsewhere

**Comment by ereslibre** [comment]
> What I'm thinking is, and this concerns the /usr/bin and /usr/lib paths from the original PR, we could add options like cdi.dynamic.nvidia.mount-usr-local, mount-usr-bin, etc. I'd set them to false by default

Just checking in; would you be fine with me adapting this PR to include those two attributes?

**Comment by SomeoneSerge** [comment]
> Just checking in; would you be fine with me adapting this PR to include those two attributes?

1. I'd support documenting "a way" of making applications that hard-code these paths (like ollama) "just work", thereby also acknowledging that it's a pattern of abuse that exists.
2. I haven't considered whether a boolean flag like `mount-usr-local` is the best way to go, but is "a way".
3. For example, however, if we moved your `mounts` from `let ... in` into `options`, these boolean flags become kind of "redundant"

Sorry, I'm not thinking clearly today and I'm not very responsive this week. What do you think about the above points?

**Comment by ereslibre** [comment]
> Sorry, I'm not thinking clearly today and I'm not very responsive this week. What do you think about the above points?

Let me experiment a little with different options. I am also not completely sold on any of them. The `mounts` option might seem very powerful because we get out of the equation to the user, however I do wonder if that is the best abstraction.

**Comment by ereslibre** [comment]
@SomeoneSerge 👋 I think this has now a good balance: it's possible to disable binary mounting and the container `/usr/local/nvidia` with `mount-nvidia-binaries` and `mount-nvidia-docker-1-directories`.

By default they are enabled so that users that enable `virtualisation.containers.cdi.dynamic.nvidia.enable` have a very complete experience out of the box, while it also allows more experienced users to mount less on containers if they wish.

I was weighing adding a list of mounts exposed to users, but I think it's better to wait for the use case to show up.

Regarding `closureinfo`, do you think I can handle that on this PR, or would you prefer to tackle that on a specific PR?

Thanks :)

**Comment by SomeoneSerge** [comment]
Hi and thanks! I'll just drop the comments:

> I was weighing adding a list of mounts exposed to users, but I think it's better to wait for the use case to show up.

1. Having to repeat the fairly long names of the boolean flags in all of `cdi-generate.nix`, `callPackage ./cdi-generate.nix`, and in `options` is, to my thinking, a use-case. I think the simplest `cdi-generate` you can make is the one parameterized by `mounts :: List { hostPath, containerPath }` (I'd also consider making it an attrset `containerPath => { hostPath, containerPath` }` because attrsets are easier to override). Then the only file "aware" of `mount-nvidia-docker-1-directories` would be the module itself, and it's there you'd derive  `mounts` from the booleans.
2. We still haven't got an escape hatch. IMO we must have one so it doesn't take a PR to nixpkgs to unbreak things. Moving `mounts` into an option is the obvious candidate

> By default they are enabled so that users that enable virtualisation.containers.cdi.dynamic.nvidia.enable have a very complete experience out of the box

1. The executables... Iirc `nvidia-ctk cdi generate` mounts them too. Seems excessive, because `PATH` should be enough, but also I think the executables have a smaller potential for breaking unrelated stuff when compared to shared libraries.
2. I still think we should not mount `{,/usr/}lib`, this is just such an odd side-effect. Instead we should verify that the hook generates a valid ld.so cache

> Regarding closureinfo, do you think I can handle that on this PR, or would you prefer to tackle that on a specific PR?

A separate PR and I also realized that `closureinfo` was maybe not the best pointer. We'll need `referencesByPopularity` or https://github.com/NixOS/nixpkgs/pull/178717

File: nixos/modules/virtualisation/containers.nix [hunk 2]
```diff
@@ -29,12 +29,30 @@ in
...
+      dynamic.nvidia = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc ''
+            Enable dynamic CDI configuration for NVidia devices by running nvidia-container-toolkit on boot.
+          '';
+        };
+
+        mount-nvidia-binaries = mkOption {
```
Comment (SomeoneSerge) [inline nixos/modules/virtualisation/containers.nix:41]: Libcuda is a binary too, maybe `executables` or `tools`

File: nixos/modules/virtualisation/containers.nix [hunk 3]
```diff
@@ -29,12 +29,30 @@ in
...
+          description = lib.mdDoc ''
+            Mount binaries nvidia-smi, nvidia-cuda-mps-control, nvidia-cuda-mps-server, nvidia-debugdump, nvidia-powerd and nvidia-ctk on containers.
+          '';
+        };
+
+        mount-nvidia-docker-1-directories = mkOption {
+          default = true;
+          type = types.bool;
+          description = lib.mdDoc ''
+            Mount nvidia-docker-1 directories on containers: /usr/local/nvidia/lib and /usr/local/nvidia/lib64.
```
Comment (SomeoneSerge) [inline nixos/modules/virtualisation/containers.nix:53]: Maybe a pointer as to why something so strange would ever be needed. E.g. I think you shared a link to that dockerfile from ollama before?

File: nixos/modules/virtualisation/containers.nix [hunk 4]
```diff
@@ -29,12 +29,30 @@ in
...
+          description = lib.mdDoc ''
+            Enable dynamic CDI configuration for NVidia devices by running nvidia-container-toolkit on boot.
+          '';
+        };
+
+        mount-nvidia-binaries = mkOption {
+          default = true;
+          type = types.bool;
+          description = lib.mdDoc ''
+            Mount binaries nvidia-smi, nvidia-cuda-mps-control, nvidia-cuda-mps-server, nvidia-debugdump, nvidia-powerd and nvidia-ctk on containers.
```
Comment (SomeoneSerge) [inline nixos/modules/virtualisation/containers.nix:45]: This isn't exact though? I think we always mount them but normally we'd mount the store path and expose them in `PATH`. What we need an explicit user's consent for is for contaminating `/usr/bin`

File: nixos/modules/virtualisation/containers.nix (hunk 2 above)
Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:41]: +1, yeah, `tools` might be a more accurate name.

File: nixos/modules/virtualisation/containers.nix (hunk 4 above)
Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: Without this setting, those executables will no tbe mounted at all on the container namespace unless I have overlooked something; I cannot find them being mounted directly nor indirectly within the container.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: Mhm. I think the right thing to do is to mount `/nix/store/.../bin/nvidia-smi=/nix/store/.../bin/nvidia-smi` and to extend `PATH`. AFAIU `containerEdits.env` lets one do that?

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: Unfortunately we cannot do that. You can set or override envvars, but cannot extend an existing one. It's not possible to do `PATH=something-new:$PATH` so to say on the OCI hook.

However, I believe the original comment was about the fact of mounting these executables and my comment was to confirm that if you disable this option, no executables will be mounted at all.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: > Unfortunately we cannot do that. You can set or override envvars, but cannot extend an existing one. It's not possible to do PATH=something-new:$PATH so to say on the OCI hook.

That's alright, we have the module system at our disposal to take care of the module sematnics. And then our script can also graciously handle the potential values `nvidia-ctk` could have set before.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: 1. We've implicitly already allocated the space for `virtualisation.containers.cdi.static."nvidia-container-toolkit".containerEdits.env`, but I'm not sure if we should make it an attrset on our side and convert to a list when passing to `cdi-generate.nix`
2. Maybe we make a separate option...

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: I don't think we can do much with `PATH`. Users might have built container images with a specific `PATH` and we should not override it, given we cannot expand it.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: Offtopic: I'm still wondering if `cdi.static` and `cdi.dynamic could have been `cdi.nvidia-container-toolkit.settings` and `cdi.nvidia-container-toolkit` respectively.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: Hmmm, populating `PATH` may still be an issue when we deal with multiple devices/multiple hooks

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: > Offtopic: I'm still wondering if cdi.static and cdi.dynamic could have been cdi.nvidia-container-toolkit.settings and cdi.nvidia-container-toolkit respectively.

- `cdi.static` gives the generic ability to configure CDI for generic devices in a manual fashion, just making sure by the type system that what they provide is a valid syntactic JSON.

- `cdi.dynamic` is a bit the "counterpart", where providers (only nvidia for now) provide you with the generic ability to configure in runtime based on the vendor detected hardware.

`cdi.static` has nothing to do with `nvidia-container-toolkit` necessarily, so I would keep them separate.

Also, having `cdi.nvidia-container-toolkit` directly could be an option, but given there is a very good split on the CDI spec of static/dynamic and their relationship (priority dynamic > static), I thought it was a good strategy to map them in that way in our options.

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: Note on that: https://github.com/cncf-tags/container-device-interface/blob/780f60f5e4ac27d7b1c729a986ed7edbb6af7c12/pkg/cdi/doc.go#L157-L164

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: > cdi.static gives the generic ability to configure
> cdi.dynamic is a bit the "counterpart",
> cdi.static has nothing to do with nvidia-container-toolkit necessarily, so I would keep them separate.

Yes, you had explained the logic in the original PR. However, we do implicitly maintain both "static" and "dynamic" options for `/run/cdi/nvidia-container-toolkit.json`: all of our `mounts` are currently computed statically (although we really should wrap them into a build with `exportReferencesGraph`), but we also do run `nvidia-ctk cdi generate` which is dynamic. And we still have to take care of merging the dynamic part with the static overrides (currently via `jq`).

I'd rather there were exactly one root option for `/run/cdi/nvidia-container-toolkit.json` (e.g. `cdi.nvidia-container-toolkit`). It is only logical to describe the static part in the style of RFC 42 ("structural `settings`"). The dynamic part then consists of running arbitrary hooks (`nvidia-ctk cdi generate`), possibly expanding the static configuration (e.g. when `mounts` contain store paths they need to be passed through the `exportReferencesGraph`), and finally merging the two

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: This is why I was suggesting that maybe we should remove `cdi.dynamic.nvidia` and keep `services.nvidia-ctk`(?):

```
options.virtualisation.containers.cdi = mkOption { type = attrsOf (submodule { options.containerEdits.env = mkOption { type = attrsOf str; }; ... }); ... } # Literally mirroring the full CDI spec, because we can
options.services.nvidia-container-toolkit.enable = ...
options.services.nvidia-container-toolkit.mount-usr-local-bin = ...
options.services.nvidia-container-toolkit.mount-usr-lib = ...
config = mkMerge [
  (mkIf config.services.nvidia-ctk.enable {
    # Corresponds to /run/cdi/nvidia-container-toolkit.json
    virtualisation.containers.cdi.nvidia-container-toolkit.containerEdits.mounts = ...
  })
]
```

etc

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:45]: > However, we do implicitly maintain both "static" and "dynamic" options for /run/cdi/nvidia-container-toolkit.json

Ideally we wouldn't have to add any mounts with `jq` on the dynamically generated JSON with `nvidia-container-toolkit cdi generate`. This is the easiest path to make this work. Ideally, we should contribute changes to  the nvidia-container-toolkit to allow for more customization when it makes sense for that project and reduce the number of patches we do on the generated JSON.

Despite we patch the generated JSON, everything we write into `/run/cdi` I consider it "dynamic" because is what goes into the "dynamically generated, although extended a bit by us to make it work fine". Without those patches, the JSON generated by the nvidia container toolkit under /var/run/cdi would not work on a NixOS host.

> I'd rather there were exactly one root option for /run/cdi/nvidia-container-toolkit.json (e.g. cdi.nvidia-container-toolkit)

I think it all boils down to the fact that you see `nvidia-container-toolkit-cdi-generator` as a module that should not exist, and instead be implemented in `virtualisation.containers` directly?

We can maybe get rid of this internal module (nvidia-container-toolkit-cdi-generator) and make the move if that is your main concern. I think it is a clean separation though.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: > Note on that: https://github.com/cncf-tags/container-device-interface/blob/780f60f5e4ac27d7b1c729a986ed7edbb6af7c12/pkg/cdi/doc.go#L157-L164

My impression is that this only concerns the generic placeholder static configs with default values that you distribute with packages in distributions like ubuntu. The "static config" we are discussing actually contain values specified by the end user, i.e. they are more relevant and informed than the values produced by `nvidia-ctk cdi generate` and should generally take the priority.

TLDR: the spec's distinction between "static" and "dynamic" so far seems fairly useless, especially given that it doesn't cover merge semantics

> Ideally, we should contribute changes to the nvidia-container-toolkit to allow for more customization when it makes sense for that project and reduce the number of patches we do on the generated JSON.

What I see so far suggests that we'll always want to apply "edits" to the configs generated by tools like "nvidia-ctk" (including those by other vendors). So I think we need a generic system for describing and applying the common overrides.

Comment (SomeoneSerge) [reply nixos/modules/virtualisation/containers.nix:45]: > I think it all boils down to the fact that you see nvidia-container-toolkit-cdi-generator as a module that should not exist, and instead be implemented in virtualisation.containers directly?

Either that or the `cdi.dynamic.nvidia` shouldn't, because they're so entangled they just mirror each other.

BUt that's not the point. The point is that ultimately we're working on a module to manage files in `/run/cdi/`. Right now there's no generic formula to say which options correspond to which files.

**Comment by ereslibre** [comment]
@SomeoneSerge Please have a look when you can. I removed the `static` and `dynamic` distinction. I think it's also not worth it for the NixOS user in the end.

It's now `hardware.nvidia-container-toolkit` only. I added a generic `mounts` attribute that is populated automatically if `mount-nvidia-executables` or `mount-nvidia-docker-1-directories` are true.

Please, let me know if something is not idiomatic or if it could be improved. Thanks :)

File: nixos/modules/virtualisation/containers.nix
Comment (SomeoneSerge) [inline nixos/modules/virtualisation/containers.nix:1]: Thanks @ereslibre, I hope to have a look tomorrow! Btw, what'd you think about jumping on a call to brainstorm the next steps? Also are you on matrix?

Comment (ereslibre) [reply nixos/modules/virtualisation/containers.nix:1]: Thank you! +1 on the call. You can find me at @ereslibre:matrix.org :)

**Comment by ereslibre** [comment]
Rebased this. I would like to include this for NixOS 24.05 (https://github.com/NixOS/nixpkgs/issues/303285). I think removing `cdi.static` and `cdi.dynamic` attributes that were included in fd464f0543b9b5370b1b3c165462e9f988d038bb and 8ba61ebb8acb51bf3d249a10a4671c5aaec7d91d.

I think this is a better balance than doing the strict separation of `cdi.static` and `cdi.dynamic` for NixOS.

**Comment by SomeoneSerge** [comment]
- > I think removing cdi.static and cdi.dynamic attributes that were included in https://github.com/NixOS/nixpkgs/commit/fd464f0543b9b5370b1b3c165462e9f988d038bb and https://github.com/NixOS/nixpkgs/commit/8ba61ebb8acb51bf3d249a10a4671c5aaec7d91d is a better balance

  👍🏻 
- I think a typed option for CDI specs is still warranted. Note that any custom specs in `etc."cdi/..."` that specify "hostPaths" will need to handle `exportReferencesGraph` anyway.
- Needs an `mkRenamedOptionModule` for from `...cdi.dynamic.nvidia` to the new service for unstable users

**Comment by ereslibre** [comment]
> I think a typed option for CDI specs is still warranted. Note that any custom specs in etc."cdi/..." that specify "hostPaths" will need to handle exportReferencesGraph anyway.

I think we can implement the typed option for CDI specs. I would suggest to do so as a follow up after this has been merged, and a way for users to specify their own CDI configurations statically that is properly type checked.

> Needs an mkRenamedOptionModule for from ...cdi.dynamic.nvidia to the new service for unstable users

Right, thanks for the reminder. Adding it later today.

**Comment by ereslibre** [comment]
> Right, thanks for the reminder. Adding it later today.

Just for the sake of notifying; this was pushed that very same day already.

cc/ @SomeoneSerge

**Comment by ereslibre** [comment]
@SomeoneSerge: thank you. Splitted in two commits as you suggested.

I will post in https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/ when it's merged.

> Notably, we still need to follow up with a fix to the incomplete deployment issue (dependencies of the host drivers and tools not being mounted).

Absolutely, we have to follow up on this. Thanks for your feedback during this process.

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:1]: My interpretation of https://github.com/NixOS/nixpkgs/blob/3b8a1a7a70b528c6dd31b160fd7ad3b32d983dbf/CONTRIBUTING.md#commit-conventions was more along the lines of "nixos/nvidia-container-toolkit: rename from virtualisation.containers.cdi" and "nixos/nvidia-container-toolkit: add mount-nvidia-binaries and ... options"

**Review by SomeoneSerge — APPROVED** [review]
If you've tested the runtime, I'm happy to merge now (I can test later)

**Comment by ereslibre** [comment]
I just re-run it, but let me do a more thorough check before finally merging, everything should be fine but will comment here. 

I will also double check the Docker issue and if it’s what I presume I will open another PR to fix that so it also makes it for 24.05.

**Comment by ereslibre** [comment]
> If you've tested the runtime, I'm happy to merge now (I can test later)

Confirmed that it's working fine here in all the cases I tested.
